frame/society/src/lib.rs

```diff
@@ -0,0 +1,652 @@
+	Vouch(AccountId, Balance),
+}
+
+impl<AccountId: PartialEq, Balance> BidKind<AccountId, Balance> {
+	fn check_voucher(&self, v: &AccountId) -> rstd::result::Result<(), &'static str> {
+		if let BidKind::Vouch(ref a, _) = self {
+			if a == v {
+				Ok(())
+			} else {
+				Err("incorrect identity")
+			}
+		} else {
+			Err("not vouched")
+		}
+	}
+}
+
+// This module's storage items.
+decl_storage! {
+	trait Store for Module<T: Trait> as Fratority {
+		/// The current bids.
+		Bids: Vec<(BalanceOf<T>, T::AccountId, BidKind<T::AccountId, BalanceOf<T>>)>;
+
+		/// The current set of members, ordered.
+		Members get(fn members)
+		// The following can be enabled once we have fixed #4315
+/*			build(|config: &GenesisConfig<T>| {
+				let mut m = config.members;
+				m.sort();
+				m
+			})*/:
+			Vec<T::AccountId>;
+
+		/// Members currently vouching, stored ordered
+		Vouching: Vec<T::AccountId>;
+
+		/// The current set of candidates; bidders that are attempting to become members.
+		pub Candidates get(candidates):
+			Vec<(BalanceOf<T>, T::AccountId, BidKind<T::AccountId, BalanceOf<T>>)>;
+
+		/// Amount of our account balance that is specifically for the next round's bid(s).
+		Pot: BalanceOf<T>;
+
+		/// The most primary from the most recently approved members.
+		Head get(head)
+		// The following can be enabled once we have fixed #4315
+		/*build(|config: &GenesisConfig<T>| config.members.first().cloned())*/:
+			Option<T::AccountId>;
+
+		/// Double map from Candidate -> Voter -> (Maybe) Vote.
+		Votes: double_map
+			hasher(twox_64_concat) T::AccountId,
+			twox_64_concat(T::AccountId)
+		=> Option<Vote>;
+
+		/// Pending payouts; ordered by block number, with the amount that should be paid out.
+		Payouts: map T::AccountId => Vec<(T::BlockNumber, BalanceOf<T>)>;
+
+		/// The ongoing number of losing votes cast by the member.
+		Strikes: map T::AccountId => StrikeCount;
+	}
+	add_extra_genesis {
+		config(members): Vec<T::AccountId>;
+	}
+}
+
+// The module's dispatchable functions.
+decl_module! {
+	/// The module declaration.
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		// Initializing events
+		// this is needed only if you are using events in your module
+		fn deposit_event() = default;
+
+		/// Make a bid for entry.
+		///
+		/// Alters Bids (O(N) decode+encode, O(logN) search, <=2*O(N) write).
+		pub fn bid(origin, value: BalanceOf<T>) -> Result {
+			let who = ensure_signed(origin)?;
+			T::Currency::reserve(&who, T::CandidateDeposit::get())?;
+
+			Self::put_bid(who.clone(), value.clone(), BidKind::Deposit);
+			Self::deposit_event(RawEvent::Bid(who, value));
+			Ok(())
+		}
+
+		pub fn unbid(origin, pos: u32) -> Result {
+			let who = ensure_signed(origin)?;
+
+			let pos = pos as usize;
+			<Bids<T>>::mutate(|b|
+				if pos < b.len() && b[pos].1 == who {
+					b.remove(pos);
+					Self::deposit_event(RawEvent::Unbid(who));
+					Ok(())
+				} else {
+					Err("bad position")
```

Comment:
This error message isn't great because it will be `bad position` if `who` is not bidder (i.e, not in the vector)

So I would suggest two different errors instead.

Reply:
meh - users shouldn't be calling into this directly so really shouldn't be an issue. if it can be done for zero extra cost, sure, but i don't think it can...

Reply:
Right, feel free to ignore me then but a comment about it would good so it won't be interpreted as a bug by future readers.

FWIW, I think rust handles `&static str` just as a "string constant" (stored in text/data segment) and it will not have any additional runtime cost (except copy it from flash to RAM at startup) but it will produce a slightly bigger binary which might be an issue for WASM.

Reply:
i mean for finding the alternative condition, not for the extra string constant.

Reply:
ahh, alright I think it would require an extra branch